src/main/scala/firrtl/passes/AnnotateMemMacros.scala

```diff
@@ -0,0 +1,167 @@
+package firrtl.passes
+
+import scala.collection.mutable.{ArrayBuffer,HashMap}
+import AnalysisUtils._
+import firrtl.WrappedExpression._
+import firrtl.ir._
+import firrtl._
+import firrtl.Utils._
+import firrtl.Mappers._
+
+case class AppendableInfo(fields: Map[String,Any]) extends Info {
+  def append(a: Map[String,Any]) = this.copy(fields = fields ++ a)
+  def append(a: (String,Any)): AppendableInfo = append(Map(a))
+  def get(f: String) = fields.get(f)
+  override def equals(b: Any) = b match {
+    case i: AppendableInfo => fields - "info" == i.fields - "info"
+    case _ => false
+  }
+}
+
+object AnalysisUtils {
+
+  def getConnects(m: Module) = {
+    val connects = HashMap[String, Expression]()
+    def getConnects(s: Statement): Unit = s match {
+      case s: Connect =>
+        connects(s.loc.serialize) = s.expr
+      case s: DefNode =>
+        connects(s.name) = s.value
+      case s: Block =>
+        s.stmts foreach getConnects
+      case _ =>
+    }
```

Comment:
Right okay. Yeah if you're on midfirrtl you don't need to worry about conditionals (meaning whens, not muxes). Regardless the code I wrote will do the right thing. And you make a good point, it's not entirely clear how to deal with this if you have multi assignment to the same thing, it seems like this really needs to be run on midfirrtl.

ValidIf is the thing that IsInvalid inside whens maps to in the same way that connections in whens map to muxes. It kind of means "ConnectIf" in that if the predicate is false just put garbage.

By unapply matching I simply mean you can match on Connect or DefNode above in two equivalent ways:
```scala
case c: Connect =>
  connects(c.loc.serialize) = c.expr
case Connect(_, loc, expr) =>
  connects(loc.serialize) = expr
```
those matches are equivalent

Finally, for s map getConnects, while it is true that on midfirrtl the only place it matters is on Blocks, if we had conditionals then it would matter there as well. You can do it on just Block if you want. Similarly, there is map for mapping on the Expression children of Statements. In that case you could do it manually (which requires a lot more than just Block), or you could use map.